### ISSUE TITLE
Breaking change: IUserTokenStore's lifetime change in Duende.AccessTokenManagement.OpenIdConnect 3.1.0

### ISSUE THREAD

**Issue by adampaquette** [DuendeSoftware/foss#82]
I was using `Duende.AccessTokenManagement.OpenIdConnect` 3.01 and everything was working then I updated to 3.1.1 and I had authorization failures. I tried 3.1.0 and discovered the problem started with this version. 

I'm not using "using IdentityModel" and the project compiles fine. 

My code using this library is basic:

```c#
public sealed class TokenStorageOidcEvents(IUserTokenStore userTokenStore) : OpenIdConnectEvents
{
    private readonly IUserTokenStore _userTokenStore = userTokenStore;

    /// <inheritdoc />
    public override async Task TokenValidated(TokenValidatedContext context)
    {
        var exp = DateTimeOffset.UtcNow.AddSeconds(
            Double.Parse(context.TokenEndpointResponse!.ExpiresIn));

        await _userTokenStore.StoreTokenAsync(context.Principal!, new UserToken
        {
            AccessToken = context.TokenEndpointResponse.AccessToken,
            AccessTokenType = context.TokenEndpointResponse.TokenType,
            RefreshToken = context.TokenEndpointResponse.RefreshToken,
            Expiration = exp,
            Scope = context.TokenEndpointResponse.Scope
        });

        await base.TokenValidated(context);
    }
}
```

```c#
public sealed class InMemoryTokenStore : IUserTokenStore
{
    private readonly ConcurrentDictionary<Guid, UserToken> _tokens = new();

    public Task<UserToken> GetTokenAsync(ClaimsPrincipal user, UserTokenRequestParameters? parameters = null)
    {
        var userId = user.GetRequiredUserId();

        return Task.FromResult(_tokens.TryGetValue(userId, out var value) ? 
            value : 
            new UserToken { Error = "not found" });
    }

    public Task StoreTokenAsync(ClaimsPrincipal user, UserToken token, UserTokenRequestParameters? parameters = null)
    {
        var userId = user.GetRequiredUserId();
        _tokens[userId] = token;
        return Task.CompletedTask;
    }

    public Task ClearTokenAsync(ClaimsPrincipal user, UserTokenRequestParameters? parameters = null)
    {
        var userId = user.GetRequiredUserId();
        _tokens.TryRemove(userId, out _);
        return Task.CompletedTask;
    }
}
```

Setup:
```c#
oidcOptions.EventsType = typeof(TokenStorageOidcEvents);
...
.AddDistributedMemoryCache()
.AddOpenIdConnectAccessTokenManagement()
.AddBlazorServerAccessTokenManagement<InMemoryTokenStore>()
.AddTransient<TokenStorageOidcEvents>()
```

For now I will stay with 3.0.1 and .NET 9.

**Comment by josephdecock** [comment]
Can you share more details about the nature of the failure? Was there an exception? Was anything logged by the application using access token management? Was anything logged by the OP?

**Comment by josephdecock** [comment]
I would also look at the versions of the System.IdentityModel and Microsoft.IdentityModel packages that you are using, possibly as transitive dependencies. It's fairly common for mixtures of versions of those libraries to cause issues.

**Comment by josephdecock** [comment]
See https://docs.duendesoftware.com/identityserver/v7/troubleshooting/wilson/ for more on troubleshooting the System.IdentityModel issues.

**Comment by adampaquette** [comment]
@josephdecock I've found the problem. The IUserTokenStore's lifetime has been changed in verison 3.1.0 from singleton to scoped. My InMemoryTokenStore needs to be a singleton to retreive cached tokens.

I've opened a PR here to support lifetime parametrization : https://github.com/DuendeSoftware/foss/pull/83

Version 3.0.1:
````c#
    public static IServiceCollection AddBlazorServerAccessTokenManagement<TTokenStore>(this IServiceCollection services)
        where TTokenStore : class, IUserTokenStore
    {
        services.AddSingleton<IUserTokenStore, TTokenStore>();
        services.AddScoped<IUserAccessor, BlazorServerUserAccessor>();
        services.AddCircuitServicesAccessor();
        services.AddHttpContextAccessor(); // For SSR

        return services;
    }
```

Version 3.1.0
````c#
    public static IServiceCollection AddBlazorServerAccessTokenManagement<TTokenStore>(this IServiceCollection services)
        where TTokenStore : class, IUserTokenStore
    {
        services.AddScoped<IUserTokenStore, TTokenStore>();
        services.AddScoped<IUserAccessor, BlazorServerUserAccessor>();
        services.AddCircuitServicesAccessor();
        services.AddHttpContextAccessor(); // For SSR

        return services;
    }
```

**Comment by josephdecock** [comment]
This was a deliberate change that I made because implementations of the user store for blazor that rely on the authentication state can't resolve the authentication state in a singleton context. 

While this is a change in behavior, from the point of view of a blazor implementation like that, this is a bug. While this does cause breakage for implementations like yours, my assumption was that in production users would store sessions more durably (e.g., in redis or a database). Storing everything in memory like this means that when your server restarts, everyone loses their session and it also means that scaling out requires server affinity. 

That said, we did run into this in our sample that implements the store like yours. The fix was to make the collection static:

```
private static ConcurrentDictionary<Guid, UserToken> _tokens = new();
```

But that's really meant to be for demos/experiments/development, not a production-worthy implementation. We could make the store registration have the optional parameter that you propose in your PR, but I'm not sure that we need it at all. I'd rather not change the API again, to try to avoid confusion. But, I'm certainly open to feedback, and I'd be really interested to hear more about your use case. Thanks!

**Comment by adampaquette** [comment]
> This was a deliberate change that I made because implementations of the user store for blazor that rely on the authentication state can't resolve the authentication state in a singleton context.
> 
> While this is a change in behavior, from the point of view of a blazor implementation like that, registering as a singleton is a bug. While this change does cause breakage for implementations like yours, my assumption was that in production users would store sessions more durably (e.g., in redis or a database). Storing everything in memory like this means that when your server restarts, everyone loses their session and it also means that scaling out requires server affinity.
> 
> That said, we did run into this in our sample that implements the store like yours. The fix was to make the collection static:
> 
> ```
> private static ConcurrentDictionary<Guid, UserToken> _tokens = new();
> ```
> 
> But that's really meant to be for demos/experiments/development, not a production-worthy implementation. We could make the store registration have the optional parameter that you propose in your PR, but I'm not sure that we need it at all. I'd rather not change the API again, to try to avoid confusion. But, I'm certainly open to feedback, and I'd be really interested to hear more about your use case. Thanks!

I agree this is not a production proof implementation. Currently if the user is considered authenticated but the Api returns a 401, I juste reload the page, provoking the reauthentification loop. Because I can get this working with a static variable I will close my PR. In the future, I will upgrade my store. Thanks
